### PR TITLE
feat: add health check for genai deployments

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

https://hpe-aiatscale.atlassian.net/browse/MLG-1464

## Test Plan

test deploying valid and invalid hashes
```
import sys

from determined.common import api
from determined.common.api import certs
from determined.deploy import healthcheck

addr = sys.argv[1]

cert = certs.Cert(noverify=True)
healthcheck.wait_for_genai_url(addr, timeout=5, cert=cert)
```

## Commentary (optional)

sample output

```
Waiting for GenAI instance to be available................................................................................................................................
Determined cluster has been deployed, but GenAI health check has failed.
For details, SSH to master instance and check /var/log/cloud-init-output.log.
```

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/deploy/healthcheck.py [hunk 1]
```diff
@@ -53,3 +53,37 @@ def wait_for_master_url(
...
+    POLL_INTERVAL = 2
+    polling = False
+    start_time = time.time()
+    GENAI_PREFIX = "/genai"
+    check_path = GENAI_PREFIX + "/api/v1/workspaces"
+
+    try:
+        while time.time() - start_time < timeout:
+            try:
+                auth = api.Authentication(master_address=master_url, cert=cert)
```
Comment (hamidzr) [inline harness/determined/deploy/healthcheck.py:72]: we don't have support for endpoints that do not require authentication at this time.

File: harness/determined/deploy/healthcheck.py [hunk 2]
```diff
@@ -53,3 +53,37 @@ def wait_for_master_url(
...
             print()
+
+
+def wait_for_genai_url(
+    master_url: str,
+    timeout: int = DEFAULT_TIMEOUT,
+    cert: Optional[certs.Cert] = None,
+) -> None:
+    POLL_INTERVAL = 2
+    polling = False
```
Comment (szewaiyuen6) [inline harness/determined/deploy/healthcheck.py:64]: I am not sure about the purpose of this flag.

Would we not want to poll until timeout or first success, whichever comes first?

File: harness/determined/deploy/healthcheck.py [hunk 3]
```diff
@@ -53,3 +53,37 @@ def wait_for_master_url(
...
+                pass
+            if not polling:
+                polling = True
+                print("Waiting for GenAI instance to be available...", end="", flush=True)
+            time.sleep(POLL_INTERVAL)
+            print(".", end="", flush=True)
+        raise MasterTimeoutExpired
+    finally:
+        if polling:
+            print()
```
Comment (szewaiyuen6) [inline harness/determined/deploy/healthcheck.py:89]: Not sure the purpose of these two lines. Some explanation would be helpful!

File: harness/determined/deploy/healthcheck.py (hunk 2 above)
Comment (hamidzr) [reply harness/determined/deploy/healthcheck.py:64]: I borrowed this from the [master health check](https://github.com/determined-ai/determined/blob/8cd8b210f36366fb917ca6870dda2eea504755e8/harness/determined/deploy/healthcheck.py#L29-L55) without rewriting the two. 

the goal of the polling flag is purely for rendering purposes to make sure we clean up and close the output line with a newline and final msg.

File: harness/determined/deploy/healthcheck.py (hunk 3 above)
Comment (hamidzr) [reply harness/determined/deploy/healthcheck.py:89]: the normal print lines we use while waiting for the check to wrap up don't output new lines to keep the loading lines limited. this closes those with a new line (if needed)'

>             print(".", end="", flush=True)

**Review by szewaiyuen6 — APPROVED** [review]
LGTM

**Pull request merged**